server/notification-providers/smtp.js

```diff
@@ -39,76 +40,80 @@ class SMTP extends NotificationProvider {
-                    serviceStatus = (heartbeatJSON["status"] === DOWN) ? "🔴 Down" : "✅ Up";
-                }
-
-                // Break replace to one by line for better readability
-                customSubject = customSubject.replace(replaceStatus, serviceStatus);
-                customSubject = customSubject.replace(replaceName, monitorName);
-                customSubject = customSubject.replace(replaceHostnameOrURL, monitorHostnameOrURL);
-
-                subject = customSubject;
+                const tpl = engine.parse(customSubject);
+                subject = await engine.render(tpl, context);
+            }
+            if (customBody !== "") {
+                const tpl = engine.parse(customBody);
+                body = await engine.render(tpl, context);
             }
-        }
-
-        let transporter = nodemailer.createTransport(config);
-
-        let bodyTextContent = msg;
-        if (heartbeatJSON) {
-            bodyTextContent = `${msg}\nTime (${heartbeatJSON["timezone"]}): ${heartbeatJSON["localDateTime"]}`;
         }
 
         // send mail with defined transport object
+        let transporter = nodemailer.createTransport(config);
         await transporter.sendMail({
             from: notification.smtpFrom,
             cc: notification.smtpCC,
             bcc: notification.smtpBCC,
             to: notification.smtpTo,
             subject: subject,
-            text: bodyTextContent,
+            text: body,
         });
 
         return "Sent Successfully.";
     }
+
+    /**
+     * Generate context for LiquidJS
+     * @param {string} msg  the message that will be included in the context
+     * @param {?object} monitorJSON Monitor details (For Up/Down/Cert-Expiry only)
+     * @param {?object} heartbeatJSON Heartbeat details (For Up/Down only)
+     * @returns {{STATUS: string, HOSTNAME_OR_URL: string, NAME: string, monitorJSON: ?object, heartbeatJSON: ?object, msg: string}} the context
+     */
+    generateContext(msg, monitorJSON, heartbeatJSON) {
+        // Let's start with dummy values to simplify code
+        let monitorName = "Monitor Name not available";
+        let monitorHostnameOrURL = "testing.hostname";
+        let serviceStatus = "⚠️ Test";
+
+        if (monitorJSON !== null) {
+            monitorName = monitorJSON["name"];
+
+            if (monitorJSON["type"] === "http" || monitorJSON["type"] === "keyword" || monitorJSON["type"] === "json-query") {
+                monitorHostnameOrURL = monitorJSON["url"];
+            } else {
+                monitorHostnameOrURL = monitorJSON["hostname"];
+            }
+        }
+
+        if (heartbeatJSON !== null) {
+            serviceStatus = (heartbeatJSON["status"] === DOWN) ? "🔴 Down" : "✅ Up";
+        }
+        return {
+            "STATUS": serviceStatus,
+            "NAME": monitorName,
+            "HOSTNAME_OR_URL": monitorHostnameOrURL,
+            monitorJSON,
+            heartbeatJSON,
+            msg,
+        };
```

Comment:
```suggestion
        return {
            // For v1 backward compatibility only
            "STATUS": serviceStatus,
            "NAME": monitorName,
            "HOSTNAME_OR_URL": monitorHostnameOrURL,
            
            // The description should introduce these tag only
            "status": serviceStatus,
            "name": monitorName,
            "hostnameOrURL": monitorHostnameOrURL,
            monitorJSON,
            heartbeatJSON,
            msg,
        };
```

Haven't tested this code. Not sure if it is right.

Since some tags are lowercase, some are uppercase, which look inconsistent to me. Is it possible to standardize them to camelCase?

Reply:
Agree, much nicer this way